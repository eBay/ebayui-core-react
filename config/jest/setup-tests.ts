import '@testing-library/jest-dom'

declare var global: typeof globalThis;

if (typeof window.URL.createObjectURL === 'undefined') {
  Object.defineProperty(window.URL, 'createObjectURL', { value: () => {} });
}
