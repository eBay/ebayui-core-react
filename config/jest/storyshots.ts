import { render } from "@testing-library/react";
import initStoryshotsAddon from "@storybook/addon-storyshots";
import type { StoryshotsOptions } from "@storybook/addon-storyshots/dist/ts3.9/api/StoryshotsOptions";

const reactTestingLibrarySerializer = {
    print: (val, serialize, indent) => serialize(val.asFragment()),
    test: (val) => val && val.hasOwnProperty("container"),
};

export function initStoryshots(config: StoryshotsOptions) {
    initStoryshotsAddon({
        ...config,
        renderer: render,
        snapshotSerializers: [reactTestingLibrarySerializer],
    });
}
