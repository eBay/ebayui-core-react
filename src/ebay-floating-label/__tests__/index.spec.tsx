import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { EbayFloatingLabel } from '../../index'
import { initStoryshots } from '../../../config/jest/storyshots'

describe('<EbayFloatingLabel>', () => {
    describe('on input change', () => {
        it('should fire an event', () => {
            const spy = jest.fn()
            const { getByLabelText } = render(
                <EbayFloatingLabel id="input" onChange={spy} label="look ma, I float" />
            )
            const input = getByLabelText('look ma, I float')
            userEvent.type(input, '123')

            expect(spy).toBeCalled()
        })
    })

    describe('on input blur', () => {
        it('should fire an event', () => {
            const spy = jest.fn()
            const { getByLabelText } = render(
                <EbayFloatingLabel id="input" onBlur={spy} label="look ma, I float" />
            )
            const input = getByLabelText('look ma, I float')
            fireEvent.blur(input)

            expect(spy).toBeCalled()
        })

        describe('and there is no value', () => {
            it('should set the inline class in the label', () => {
                const { container, getByLabelText } = render(
                    <EbayFloatingLabel id="input" label="Label" />
                )
                const input = getByLabelText('Label')
                fireEvent.blur(input)

                expect(container.querySelector('.floating-label__label--inline')).toBeInTheDocument()
                expect(container.querySelector('.floating-label__label--animate')).toBeInTheDocument()
            })
        })
    })

    describe('on input focus', () => {
        it('should fire an event', () => {
            const spy = jest.fn()
            const { getByLabelText } = render(
                <EbayFloatingLabel id="input" onFocus={spy} label="look ma, I float" />
            )

            const input = getByLabelText('look ma, I float')
            fireEvent.focus(input)

            expect(spy).toBeCalled()
        })

        it('should animate the label', () => {
            const { container, getByLabelText } = render(
                <EbayFloatingLabel id="input" label="Label" />
            )

            const input = getByLabelText('Label')
            fireEvent.focus(input)

            expect(container.querySelector('.floating-label__label--animate')).toBeInTheDocument()
            expect(container.querySelector('.floating-label__label--inline')).not.toBeInTheDocument()
        })
    })

    describe('on passing a ref', () => {
        it('should set the ref as the input reference', () => {
            const ref = React.createRef() as any
            render(<EbayFloatingLabel ref={ref} label="look ma, I float" />)

            expect(ref.current.tagName).toBe('INPUT')
        })
    })

    it(`should show placeholder on focus only`, () => {
        const { getByLabelText } = render(
            <EbayFloatingLabel id="input" placeholder="Placeholder" label="look ma, I float" />
        )
        const input = getByLabelText('look ma, I float')
        fireEvent.blur(input)
        expect(input).not.toHaveAttribute('placeholder')

        fireEvent.focus(input)
        expect(input).toHaveAttribute('placeholder', 'Placeholder')
    })


    it(`should animate the label on changing the value`, () => {
        const { container, rerender } = render(
            <EbayFloatingLabel label="Label" />
        )

        rerender(
            <EbayFloatingLabel label="Label" value="234" />
        )

        expect(container.querySelector('.floating-label__label--inline')).not.toBeInTheDocument()
    })
})

global.getComputedStyle = jest.fn().mockReturnValue({ backgroundColor: 'rgb(255, 255, 255)' })

initStoryshots({
    config: ({ configure }) =>
        configure(() => {
            require('./index.stories')
        }, module)
})
