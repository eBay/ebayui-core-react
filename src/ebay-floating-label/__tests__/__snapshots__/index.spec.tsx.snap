// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots ebay-floating-label Autofocus test 1`] = `
<DocumentFragment>
  <div>
    <span
      class="floating-label"
    >
      <label
        class="floating-label__label floating-label__label--animate"
      >
        Autofocused field
      </label>
      <span
        class="textbox"
      >
        <input
          class="textbox__control"
          type="text"
          value=""
        />
      </span>
    </span>
    <br />
    <span
      class="floating-label"
    >
      <label
        class="floating-label__label"
      >
        Regular field
      </label>
      <span
        class="textbox"
      >
        <input
          class="textbox__control"
          type="text"
          value=""
        />
      </span>
    </span>
  </div>
</DocumentFragment>
`;

exports[`Storyshots ebay-floating-label Control value from outside 1`] = `
<DocumentFragment>
  <p>
    <span
      class="floating-label"
    >
      <label
        class="floating-label__label"
      >
        Will convert to lowercase
      </label>
      <span
        class="textbox"
      >
        <input
          class="textbox__control"
          placeholder="Enter some UPPERCASE"
          size="30"
          type="text"
          value=""
        />
      </span>
    </span>
  </p>
  <p>
    <button
      class="btn btn--secondary"
    >
      Change text
    </button>
  </p>
  <p>
    <button
      class="btn btn--secondary"
    >
      Clear
    </button>
  </p>
</DocumentFragment>
`;

exports[`Storyshots ebay-floating-label Default floating label 1`] = `
<DocumentFragment>
  <form>
    <p>
      <span
        class="floating-label"
      >
        <label
          class="floating-label__label"
          for="input-1"
        >
          Email address
        </label>
        <span
          class="textbox"
        >
          <input
            class="textbox__control"
            id="input-1"
            type="email"
            value=""
          />
        </span>
      </span>
    </p>
    <p>
      <button
        class="btn btn--secondary"
        type="submit"
      >
        Submit
      </button>
    </p>
  </form>
</DocumentFragment>
`;

exports[`Storyshots ebay-floating-label Floating label disabled 1`] = `
<DocumentFragment>
  <span
    class="floating-label"
  >
    <label
      class="floating-label__label floating-label__label--disabled"
      for="input-4"
    >
      Email address
    </label>
    <span
      class="textbox"
    >
      <input
        class="textbox__control"
        disabled=""
        id="input-4"
        type="text"
        value=""
      />
    </span>
  </span>
</DocumentFragment>
`;

exports[`Storyshots ebay-floating-label Floating label invalid 1`] = `
<DocumentFragment>
  <span
    class="floating-label"
  >
    <label
      class="floating-label__label floating-label__label--invalid"
      for="input-5"
    >
      Email address
    </label>
    <span
      class="textbox"
    >
      <input
        aria-invalid="true"
        class="textbox__control"
        id="input-5"
        type="text"
        value=""
      />
    </span>
  </span>
</DocumentFragment>
`;

exports[`Storyshots ebay-floating-label Floating label with controlled value 1`] = `
<DocumentFragment>
  <span
    class="floating-label"
  >
    <label
      class="floating-label__label"
      for="input-21"
    >
      Email address
    </label>
    <span
      class="textbox"
    >
      <input
        class="textbox__control"
        id="input-21"
        type="text"
        value="can't change"
      />
    </span>
  </span>
</DocumentFragment>
`;

exports[`Storyshots ebay-floating-label Floating label with placeholder 1`] = `
<DocumentFragment>
  <span
    class="floating-label"
  >
    <label
      class="floating-label__label"
      for="input-3"
    >
      Email address
    </label>
    <span
      class="textbox"
    >
      <input
        class="textbox__control"
        id="input-3"
        placeholder="placeholder text"
        type="text"
        value=""
      />
    </span>
  </span>
</DocumentFragment>
`;

exports[`Storyshots ebay-floating-label Floating label with value 1`] = `
<DocumentFragment>
  <span
    class="floating-label"
  >
    <label
      class="floating-label__label"
      for="input-2"
    >
      Email address
    </label>
    <span
      class="textbox"
    >
      <input
        class="textbox__control"
        id="input-2"
        type="text"
        value="test"
      />
    </span>
  </span>
</DocumentFragment>
`;

exports[`Storyshots ebay-floating-label Large Floating label 1`] = `
<DocumentFragment>
  <span
    class="floating-label floating-label--large"
  >
    <label
      class="floating-label__label"
      for="input-2"
    >
      Email address
    </label>
    <span
      class="textbox"
    >
      <input
        class="textbox__control textbox__control--large"
        id="input-2"
        type="text"
        value="test"
      />
    </span>
  </span>
</DocumentFragment>
`;

exports[`Storyshots ebay-floating-label Select Floating label 1`] = `
<DocumentFragment>
  <span
    class="floating-label"
  >
    <label
      class="floating-label__label"
      for="input-2"
    >
      State
    </label>
    <span
      class="select"
    >
      <select
        id="input-2"
      >
        <option
          value=""
        />
        <option
          value="CA"
        >
          California
        </option>
        <option
          value="NY"
        >
          New York
        </option>
      </select>
      <svg
        aria-hidden="true"
        class="icon icon--dropdown"
        focusable="false"
        height="8"
        width="8"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlink:href="#icon-dropdown"
        />
      </svg>
    </span>
  </span>
</DocumentFragment>
`;
